### PR TITLE
[CAS depscan] Ensure `CASOptions` are passed via `CompilerInvocation` and remove `extractCASFromFS`

### PR DESCRIPTION
**Pull request by akyrtzi**
Use `DependencyScanningService` to hold and pass `CASOptions`, which simplifies APIs since it avoids the need to be passing it around across APIs.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: clang/lib/Frontend/CompilerInstance.cpp [hunk 1]
```diff
@@ -881,15 +846,6 @@ llvm::cas::CASDB &CompilerInstance::getOrCreateCAS() {
...
 
-  // Return the CAS from the VFS, if any.
-  //
-  // FIXME: This layering is hard to follow; instead, change clients to set the
-  // CAS when building a VFS with a CAS not available from CompilerInvocation.
-  if (FileMgr)
-    if (llvm::cas::CASDB *FSCAS =
-            extractCASFromFS(FileMgr->getVirtualFileSystem()))
-      return *FSCAS;
-
```
Comment (dexonsmith) [inline clang/lib/Frontend/CompilerInstance.cpp:892]: I know I wrote this twisted code, but I'm having some trouble figuring out exactly why, or why it's safe to delete it all now.

Can you explain why it's okay?

(Maybe it predates getOrCreateCAS() existing, and should have been deleted then?)

Comment (akyrtzi) [reply clang/lib/Frontend/CompilerInstance.cpp:892]: Maybe it was more useful before the consolidation of `CASOptions`? It is easier now for clients to setup a CAS and pass it along.

Comment (dexonsmith) [reply clang/lib/Frontend/CompilerInstance.cpp:892]: Looking at git-blame, this was all added at the same time that `CASOptions` was consolidated: 112f361cfb244a209a38f5f2bdb53e4a1b818331.

File: clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp [hunk 2]
```diff
@@ -142,16 +142,17 @@ class DependencyScanningAction : public tooling::ToolAction {
...
       llvm::IntrusiveRefCntPtr<DependencyScanningCASFilesystem> DepCASFS,
       ExcludedPreprocessorDirectiveSkipMapping *PPSkipMappings,
       bool OverrideCASTokenCache, ScanningOutputFormat Format,
       bool OptimizeArgs, bool EmitDependencyFile,
       llvm::Optional<StringRef> ModuleName = None)
       : WorkingDirectory(WorkingDirectory), Consumer(Consumer),
-        DepFS(std::move(DepFS)), DepCASFS(std::move(DepCASFS)),
-        PPSkipMappings(PPSkipMappings), Format(Format),
-        OverrideCASTokenCache(OverrideCASTokenCache),
+        CASOpts(CASOpts), DepFS(std::move(DepFS)),
```
Comment (dexonsmith) [inline clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:153]: Seems like `std::move(CASOpts)` would be better.

File: clang/lib/Frontend/CompilerInstance.cpp [hunk 3]
```diff
@@ -160,42 +160,7 @@ llvm::vfs::FileSystem &CompilerInstance::getVirtualFileSystem() const {
...
-static llvm::cas::CASDB *extractCASFromFS(const llvm::vfs::FileSystem &FS) {
-  if (!FS.isCASFS())
-    return nullptr;
-  return &static_cast<const llvm::cas::CASFileSystemBase &>(FS).getCAS();
-}
-
-void CompilerInstance::setFileManager(FileManager *Value) {
-  /// FIXME: Stop checking this here; instead, trust anyone that overrides the
-  /// file manager to also override the CAS when appropriate. Needs some
-  /// layering fixes in clang-scan-deps though.
```
Comment (dexonsmith) [inline clang/lib/Frontend/CompilerInstance.cpp:172]: I believe this FIXME is the key, and your patch is addressing it.

File: clang/lib/Frontend/CompilerInstance.cpp (hunk 1 above)
Comment (dexonsmith) [reply clang/lib/Frontend/CompilerInstance.cpp:892]: I've convinced myself that the patch is correct. The FIXME I just added a comment to in the deleted code in `setFileManager()` was apparently a future which this patch is executing :).

**Review by dexonsmith — APPROVED** [review]
LGTM, with one inline suggestion.

File: clang/include/clang/Tooling/DependencyScanning/DependencyScanningWorker.h [hunk 4]
```diff
@@ -111,6 +112,7 @@ class DependencyScanningWorker {
   /// Whether to optimize the modules' command-line arguments.
   bool OptimizeArgs;
 
+  const CASOptions &CASOpts;
```
Comment (benlangmuir) [inline clang/include/clang/Tooling/DependencyScanning/DependencyScanningWorker.h]: I would suggest copying the value or making it a `shared_ptr` if you think it's too expensive to copy.  The libclang documentation says the service should outlive the workers, but it's not obvious to me that it shouldn't be allowed to go away earlier, and if someone gets it wrong it's going to be UAF potentially deep inside the worker, and might be hard to diagnose.

File: clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp (hunk 2 above)
Comment (akyrtzi) [reply clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:153]: Do you mean that I should change `DependencyScanningAction.CASOpts` to be a copy (`CASOptions CASOpts;`) instead of a reference (`const CASOptions &CASOpts;`)?

Note that `DependencyScanningAction` is not self-contained, there are these fields:
```
  DependencyConsumer &Consumer;
  ExcludedPreprocessorDirectiveSkipMapping *PPSkipMappings;
```
which indicates that `DependencyScanningAction` should not outlive `DependencyScanningWorker` or its caller, so there's no gain to avoid the reference, AFAICT.

File: clang/include/clang/Tooling/DependencyScanning/DependencyScanningWorker.h (hunk 4 above)
Comment (akyrtzi) [reply clang/include/clang/Tooling/DependencyScanning/DependencyScanningWorker.h]: I've changed it to a copy.👍

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Review by benlangmuir — APPROVED** [review]

File: clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp (hunk 2 above)
Comment (akyrtzi) [reply clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:153]: @dexonsmith I'll go ahead and merge but let me know if you think I should make a change here and I'll follow-up.

Comment (dexonsmith) [reply clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:153]: I must have misread this, and thought this was a constructor for one of the fields where it *was* a copy. 141d745a476f8093fad48bf7e9809d3fe6dc8289 LGTM!
